tests/flashcore/test_database.py

```
@@ -0,0 +1,532 @@
+        retrieved = db.get_card_by_uuid(card.uuid)
+        assert isinstance(retrieved.tags, set) and "tagX" in retrieved.tags
+        assert isinstance(retrieved.media_paths, list) and Path("foo.png") in retrieved.media_paths
+        assert isinstance(retrieved.source_yaml_file, Path) and retrieved.source_yaml_file.name == "foo.yaml"
+
+    def test_upsert_cards_batch_transaction_rollback(self, db_manager: FlashcardDatabase):
+        db = db_manager
+        db.initialize_schema()
+        valid_card = create_sample_card()
+        # Intentionally break a card (deck_name=None violates NOT NULL)
+        broken_card = create_sample_card(deck_name=None)
+        with pytest.raises(Exception):
```

Ruff (0.11.9)

Do not assert blind exception: `Exception`

(B017)